pkg/util/validation/exporter/exporter_test.go

```diff
@@ -101,3 +114,82 @@ cortex_limits_defaults{limit_name="ruler_max_rule_groups_per_tenant"} 32
 	err = testutil.CollectAndCompare(exporter, bytes.NewBufferString(limitsMetrics), "cortex_limits_defaults")
 	assert.NoError(t, err)
 }
+
+func TestOverridesExporterWithRing(t *testing.T) {
+	tenantLimits := map[string]*validation.Limits{
+		"tenant-a": {},
+	}
+
+	ringStore, closer := consul.NewInMemoryClient(ring.GetCodec(), log.NewNopLogger(), nil)
+	t.Cleanup(func() { assert.NoError(t, closer.Close()) })
+
+	cfg := Config{RingConfig{Enabled: true}}
+	cfg.Ring.KVStore.Mock = ringStore
+	cfg.Ring.InstanceID = "overrides-exporter"
+	cfg.Ring.InstanceAddr = "1.2.3.4"
+	cfg.Ring.InstancePort = 1234
+
+	// Create an empty ring.
+	ctx := context.Background()
+	require.NoError(t, ringStore.CAS(ctx, ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		return ring.NewDesc(), true, nil
+	}))
+
+	exporter, err := NewOverridesExporter(cfg, &validation.Limits{}, validation.NewMockTenantLimits(tenantLimits), log.NewNopLogger(), nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartManagerAndAwaitHealthy(ctx, exporter.subserviceManager))
+	t.Cleanup(func() { require.NoError(t, services.StopManagerAndAwaitStopped(ctx, exporter.subserviceManager)) })
+
+	// Register this instance in the ring.
+	require.NoError(t, ringStore.CAS(ctx, ringKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		desc := in.(*ring.Desc)
+		desc.AddIngester(
+			cfg.Ring.InstanceID,
+			fmt.Sprintf("%s:%d", cfg.Ring.InstanceAddr, cfg.Ring.InstancePort),
+			"",
+			[]uint32{1},
+			ring.ACTIVE,
+			time.Now(),
+		)
+		return desc, true, nil
+	}))
+
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Why is this required? Using `time.Sleep()` could lead to flaky tests. Instead, if you need to wait on a condition, please use `test.Poll()`.

Reply:
I think here you want to wait until the ring client got the updated state. You can directly use the `exporter`'s ring client and poll it.

Reply:
thanks for the hint, fixed here and below